protozfitsreader/__init__.py

```diff
@@ -1,262 +1,313 @@
 #!/bin/env python
-# this code should run in python3.
-# Zfits/protobuf loader.
-# import protozfitsreader
+from os.path import isfile
 import numpy as np
 from . import rawzfitsreader
 from . import L0_pb2
+import warnings
+
+pixel_remap = [
```

Comment:
I suppose pixel_remap and the argsort you do are the same. If they are then we can remove pixel_remap

Reply:
or better transform this as a test.

```python
def test_pixel_mapping():
    from protozfitsreader import ZFile
        for event in ZFile(example_file_path):
            assert event._sort_ids == pixel_remap 
            # some warning saying you probably have the wrong pixel mapping or it has been changed.




```

Reply:
This is protozfitsreader.py as it was in October 25 2017, when I started working on SST1M.
https://github.com/cta-sst-1m/digicampipe/blob/2df1440705328e1a9886018d59959235eb3be454/digicampipe/io/protozfitsreader.py

This:
https://github.com/cta-sst-1m/digicampipe/blob/2df1440705328e1a9886018d59959235eb3be454/digicampipe/io/protozfitsreader.py#L291-L292

is a sort of "argsort" which uses a feature in Py3.5 that dict keys happen to be sorted. In Py3.6. dict keys were not sorted anymore, so this part was replaced by np.argsort. 

I personally have never seen `pixel_remap` being used for anything.

Reply:
pixel_remap was used for some events that were not recorded with the correct pixel mapping. 
i.e. the ZfitsWriter/CameraServer was not started with a mapping file. This is now impossible since if you start without a mapping file the CameraServer will exit. 
So I suppose the pixel_remap is actually the order from which pixels are "normally" arriving. Therefore we might have : 
```python
assert event._sort_ids == np.arange(0, event._sort_ids.shape[0]) 
```
the reader could check if a mapping (at first iteration) is applied: 

```python
if event._sort_ids == pixel_remap: 
   print('pixel mapping was probably not used in file writer')
```

I am not sure if it is actually the opposite of what I am saying.

Reply:
Neither:
```python
    from digicampipe.io.protozfitsreader import pixel_remap
    assert (event._sort_ids == pixel_remap).all()
```

nor this
```python
assert (event._sort_ids == np.arange(0, event._sort_ids.shape[0])).all()
```
passes.

sorter_ids, looks like this:
```
[332 961 333 328 326 963 960 967 334 329 327 308 973 962 965 966]
[335 330 324 309 304 302 975 972 979 964 970 971 320 331 325 310]
[305 303 188 913 974 977 978 968 969 985 321 316 314 311 306 300]
[189 184 182 915 912 919 976 982 983 987 984 991 322 317 315 296]
[307 301 190 185 183 164 925 914 917 918 980 981 997 986 989 990]
[323 318 312 297 292 290 191 186 180 165 160 158 927 924 931 916]
[ 922  923  999  996 1003  988  994  995  380  319  313  298  293  291
  176  187]
[ 181  166  161  159   44  865  926  929  930  920  921  937  998 1001
 1002  992]
[ 993 1105  381  376  374  299  294  288  177  172  170  167  162  156
   45   40]
[  38  867  864  871  928  934  935  939  936  943 1000 1006 1007 1107
 1104 1111]
[382 377 375 356 295 289 178 173 171 152 163 157  46  41  39  20]
[ 877  866  869  870  932  933  949  938  941  942 1004 1005 1117 1106
 1109 1110]
[383 378 372 357 352 350 179 174 168 153 148 146  47  42  36  21]
[  16   14  879  876  883  868  874  875  951  948  955  940  946  947
 1119 1116]
[1123 1108 1114 1115  368  379  373  358  353  351  236  175  169  154
  149  147]
[ 32  43  37  22  17  15 878 881 882 872 873 889 950 953 954 944]
[ 945 1057 1118 1121 1122 1112 1113 1129  369  364  362  359  354  348
  237  232]
[230 155 150 144  33  28  26  23  18  12 880 886 887 891 888 895]
[ 952  958  959 1059 1056 1063 1120 1126 1127 1131 1128 1135  370  365
  363  344]
[355 349 238 233 231 212 151 145  34  29  27   8  19  13 884 885]
[ 901  890  893  894  956  957 1069 1058 1061 1062 1124 1125 1141 1130
 1133 1134]
[371 366 360 345 340 338 239 234 228 213 208 206  35  30  24   9]
[   4    2  903  900  907  892  898  899 1071 1068 1075 1060 1066 1067
 1143 1140]
[1147 1132 1138 1139  428  367  361  346  341  339  224  235  229  214
  209  207]
[  92   31   25   10    5    3  902  905  906  896  897 1009 1070 1073
 1074 1064]
[1065 1081 1142 1145 1146 1136 1137 1249  429  424  422  347  342  336
  225  220]
[ 218  215  210  204   93   88   86   11    6    0  904  910  911 1011
 1008 1015]
[1072 1078 1079 1083 1080 1087 1144 1150 1151 1251 1248 1255  430  425
  423  404]
[343 337 226 221 219 200 211 205  94  89  87  68   7   1 908 909]
[1021 1010 1013 1014 1076 1077 1093 1082 1085 1086 1148 1149 1261 1250
 1253 1254]
[431 426 420 405 400 398 227 222 216 201 196 194  95  90  84  69]
[  64   62 1023 1020 1027 1012 1018 1019 1095 1092 1099 1084 1090 1091
 1263 1260]
[1267 1252 1258 1259  416  427  421  406  401  399  284  223  217  202
  197  195]
[  80   91   85   70   65   63 1022 1025 1026 1016 1017 1033 1094 1097
 1098 1088]
[1089 1201 1262 1265 1266 1256 1257 1273  417  412  410  407  402  396
  285  280]
[ 278  203  198  192   81   76   74   71   66   60 1024 1030 1031 1035
 1032 1039]
[1096 1102 1103 1203 1200 1207 1264 1270 1271 1275 1272 1279  418  413
  411  392]
[ 403  397  286  281  279  260  199  193   82   77   75   56   67   61
 1028 1029]
[1045 1034 1037 1038 1100 1101 1213 1202 1205 1206 1268 1269 1285 1274
 1277 1278]
[419 414 408 393 388 386 287 282 276 261 256 254  83  78  72  57]
[  52   50 1047 1044 1051 1036 1042 1043 1215 1212 1219 1204 1210 1211
 1287 1284]
[1291 1276 1282 1283  851  415  409  394  389  387  272  283  277  262
  257  255]
[ 140   79   73   58   53   51 1046 1049 1050 1040 1041 1153 1214 1217
 1218 1208]
[1209 1225 1286 1289 1290 1280 1281  847  846  850  849  395  390  384
  273  268]
[ 266  263  258  252  141  136  134   59   54   48 1048 1054 1055 1155
 1152 1159]
[1216 1222 1223 1227 1224 1231 1288 1294 1295  827  841  840  845  844
  848  863]
[ 391  385  274  269  267  248  259  253  142  137  135  116   55   49
 1052 1053]
[1165 1154 1157 1158 1220 1221 1237 1226 1229 1230 1292 1293  823  822
  826  825]
[843 842 859 858 862 861 275 270 264 249 244 242 143 138 132 117]
[ 112  110 1167 1164 1171 1156 1162 1163 1239 1236 1243 1228 1234 1235
  707  817]
[816 821 820 824 839 853 852 857 856 860 803 271 265 250 245 243]
[ 128  139  133  118  113  111 1166 1169 1170 1160 1161 1177 1238 1241
 1242 1232]
[1233  703  702  706  705  819  818  835  834  838  837  855  854  799
  798  802]
[ 801  251  246  240  129  124  122  119  114  108 1168 1174 1175 1179
 1176 1183]
[1240 1246 1247  683  697  696  701  700  704  719  829  828  833  832
  836  779]
[ 793  792  797  796  800  815  247  241  130  125  123  104  115  109
 1172 1173]
[1189 1178 1181 1182 1244 1245  679  678  682  681  699  698  715  714
  718  717]
[831 830 775 774 778 777 795 794 811 810 814 813 131 126 120 105]
[ 100   98 1191 1188 1195 1180 1186 1187  563  673  672  677  676  680
  695  709]
[708 713 712 716 659 769 768 773 772 776 791 805 804 809 808 812]
[ 755  127  121  106  101   99 1190 1193 1194 1184 1185  559  558  562
  561  675]
[674 691 690 694 693 711 710 655 654 658 657 771 770 787 786 790]
[ 789  807  806  751  750  754  753  107  102   96 1192 1198 1199  539
  553  552]
[557 556 560 575 685 684 689 688 692 635 649 648 653 652 656 671]
[ 781  780  785  784  788  731  745  744  749  748  752  767  103   97
 1196 1197]
[535 534 538 537 555 554 571 570 574 573 687 686 631 630 634 633]
[651 650 667 666 670 669 783 782 727 726 730 729 747 746 763 762]
[766 765 529 528 533 532 536 551 565 564 569 568 572 515 625 624]
[629 628 632 647 661 660 665 664 668 611 721 720 725 724 728 743]
[757 756 761 760 764 531 530 547 546 550 549 567 566 511 510 514]
[513 627 626 643 642 646 645 663 662 607 606 610 609 723 722 739]
[738 742 741 759 758 541 540 545 544 548 491 505 504 509 508 512]
[527 637 636 641 640 644 587 601 600 605 604 608 623 733 732 737]
[736 740 543 542 487 486 490 489 507 506 523 522 526 525 639 638]
[583 582 586 585 603 602 619 618 622 621 735 734 481 480 485 484]
[488 503 517 516 521 520 524 467 577 576 581 580 584 599 613 612]
[617 616 620 483 482 499 498 502 501 519 518 463 462 466 465 579]
[578 595 594 598 597 615 614 493 492 497 496 500 443 457 456 461]
[460 464 479 589 588 593 592 596 495 494 439 438 442 441 459 458]
[475 474 478 477 591 590 433 432 437 436 440 455 469 468 473 472]
[476 435 434 451 450 454 453 471 470 445 444 449 448 452 447 446]
```

Reply:
Ok no idea what it is... Could you print the shape ?

Reply:
oh ..

Reply:
I did not see this reply

Reply:
The shape is simply 1296. I just printed it in groups of 16 so that you can see it completely

Reply:
ah ok... I was confused :D

Reply:
Yes sorry, that was my fault. It would be better I think you printed it yourself.

So the problem is not its shape, it is its contensts. You proposed two tests. and **both fail**.

Reply:
then the test should be : 

```python
assert (event._sort_ids != pixel_remap).all()
```

this would mean that the file has a different mapping than the usual one. anyway I suppose that it does not affect the data order since we read the mapping from the zfits. But a warning at __init__() could be nice.

Reply:
[I deleted the confusing information and our discussion about it, since it does not add valuable information about what `pixel_remap` is.]
So the problem is not its shape, it is its contents. You proposed two tests. and **both fail**.

Reply:
Ah now I understand, you propose to introduce a **new feature** into the reader. 
The story for this feature goes roughly like this:

Currently the adc samples of our 1296 pixels in our zfits files are written in some specific order. This order is also stored in the file, so the reader can (and does) return the adc samples not in the order it is written into the file, but in the order of pixel_ids.

So the user of the reader does not know in which order the adc samples are stored. 

You propose to inform the user, if the order in the file changes with respect to the order which was used e.g. in the 100evts. example file delivered with digicam pipe. 

Correct?

Reply:
> So the user of the reader does not know in which order the adc samples are stored.
...
> Correct?

Yes. thanks for the clarification.